walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java

```diff
@@ -270,5 +274,151 @@ public void notEnoughRoom() {
+            Arrays.asList(PERSON_B)));
+
+    MeetingRequest request = new MeetingRequest(Arrays.asList(PERSON_A), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void noMandatoryWithGaps() {
+    // Have one person, but make it so that there is just enough room at one point in the day to
+    // have the meeting.
+    //
+    // Optional: |--A--|       |----A----|              |------A------|
+    // Optional: |--B--|           |--B--|              |------B------|
+    // Day     : |---------------------------------------------------------------|
+    // Options :       |-----|           |--------------|             |----------|
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0830AM, false),
+            Arrays.asList(PERSON_A, PERSON_B)),
+        new Event("Event 2", TimeRange.fromStartEnd(TIME_0900AM, TIME_1000AM, false),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 3", TimeRange.fromStartEnd(TIME_0930AM, TIME_1000AM, false),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 4", TimeRange.fromStartEnd(TIME_1200PM, TIME_0130PM, false),
+            Arrays.asList(PERSON_A, PERSON_B)));
+
+    MeetingRequest request = new MeetingRequest(Arrays.asList(), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES), 
+            TimeRange.fromStartEnd(TIME_1000AM, TIME_1200PM, false), 
+            TimeRange.fromStartEnd(TIME_0130PM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void noMandatoryWithoutGaps() {
+    // Have one person, but make it so that there is just enough room at one point in the day to
+    // have the meeting.
+    //
+    // Optional: |----A----|
+    // Optional: |----B----|
+    // Day     : |---------|
+    // Options :      
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_A, PERSON_B)));
+
+    MeetingRequest request = new MeetingRequest(Collections.<String>emptyList(), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList();
+
+    Assert.assertEquals(expected, actual);
+  }
+}
```

Comment:
Thank you!!
As for other tests, the starter code for the assignment actually came with a bunch of tests already done. (This also means that I can't take credit for the awesome visual comments you mentioned above--the template for them was there when I arrived :)) I think 1i, 2, and 3 were included in the tests we started with that I didn't modify. 
I've got a question about 1ii: if there's an optional person who isn't in the events list, does that just mean their whole day is open?
And for 4, I'll add this if I have time to make that new branch/functionality as I described above, if that sound good!

Reply:
For 1ii: It's ambiguous! You make the judgment call hehe. Since this is a real thing that can happen, it's worth actually specifying it through tests and in the docs, since users of this code might have the same question, and might even build something the assumes one behavior or the other.

Reply:
For 4 - if you like; again, it's not a priority, only do it if you're learning useful things by doing it. If it's just busy work, then focus on the capstone.